Feature: Zoom

AKMaily opened the pull request:
## Checklist
Make sure you

- [x] have read the [contribution guidelines](../CONTRIBUTION.md),
- [x] given this PR a concise and imperative title, <!-- e.g. "Fix memory leak in DeviceService" or "Add dark-mode toggle" -->
- [x] have added necessary unit/e2e tests if necessary,
- [x] have added documentation if necessary,
- [x] have documented the changes in the [CHANGELOG.md](../CHANGELOG.md).

## 🛠 Type of change
<!-- Tick **all that apply** and delete the unused ones. -->

- [x] New feature


## Summary
<!-- A description of 1–3 sentences of what this PR does and *why*
What problem does this PR solve?
Which concept, bug, or requirement does it address? -->

This PR implements the zoom feature for the graph.
With the proposed changes the user can zoom with the mouse wheel in and out of the graph, as well as navigating through the graph per mouse click and drag. 


Issue number #112

## 📝 Design Decisions
<!-- Changes in detail (files, concepts)
>Describe the way your implementation works or what design decisions you made if applicable.
>Which are the main files and concepts you changed or introduced? -->

The zoom behavior is initially used from the (d3-zoom documentation)[https://d3js.org/d3-zoom], especially from the first example used : https://observablehq.com/@d3/zoomable-scatterplot . 
However a few additionally changes needed to be made to fit the current graph implementation. 

1. Split responsibility
Instead of using everything in one function triggering the zoom and updating the Scales needed to be seperated in the graph component and data-service. 

2. Define types 
Instead of using the zoom object directly Zoombehavior is used to define the type first.  

3. Reactive Scale Updates 
Instead of using the zoomed function directly to rescale the axis, the zoom function only triggers the zoom to update in the data-service when a zoom behavior like scrolling the mouse wheel is used.
 Using the baseAxis Scales and zoomFactor to calculate the new AxisScale makes sure not only the values on the axis changes but also the axis themselves as well as the curve scale with the new changes. 

4. Clip Path 
Additionally set a clip path in the graph component keeping the path between the axis when zooming . 

## How to test

### 1. Expected behavior
Zooming in and out of the graph is now possible. 
When zooming out the difference between the datapoints on the x-axis and y-axis should descrease dynamically. The curve should be adjusted dynamically to the axis. 
When zooming in the difference between the datapoints on the x-axis and y-axis should increase dynamically. The curve should be adjusted dynamically to the axis. 
The curve should not be drawn outside the axis. 

### 2. Steps to reproduce
Start the application in angular. 
Start the random data server. 
Scroll with you mousewheel in and out . 
Drag the curve by clicking with you left mouse into the graph and drag.

### 3. Is there still unexpected behaviour which needs to be addressed in the future?

The relative time values generated by the Random Data Server become incorrect when zooming in or out, whereas the absolute timestamps remain stable and display as expected. This needs further investigation. A related inaccuracy also appears when computing the relative axis for OmnAIScope data, so I’ll open a dedicated issue to track both problems.

AKMaily commented on angular-frontend/src/app/graph/graph-data.service.ts:
```diff
@@ -1,17 +1,31 @@
 type xDomainTuple = [xDomainType, xDomainType];
 
-const defaultXDomain: xDomainTuple = [new Date(), new Date(Date.now() - 24 * 60 * 60 * 1000)];
+const defaultXDomain: xDomainTuple = [
+  new Date(),
+  new Date(Date.now() - 24 * 60 * 60 * 1000),
+];
```
The CI currently does not have a formatter as those PRs were not merged yet. I kept them open to allow students to still work on this project without having a lot of merge issues. 

I ran prettier over the code but I need to check if it has the same settings as discussed.

tabeatheunicorn commented on angular-frontend/src/app/graph/graph-data.service.ts:
```diff
@@ -33,56 +49,72 @@ export class DataSourceService {
 
+  private readonly $zoom = signal<ZoomTransform>(zoomIdentity);
+
+  setZoom(t: ZoomTransform) {
+    this.$zoom.set(t);
+  }
```
As for the signals: I am not to sure if that makes more sense than deliberately updating the relevant other signals in that function as the zoom behavior is, like the mouse position, not really a "state" that one should keep.
While I see that this works, coming from an 'easy to understand the intention of the code' standpoint, the zoom behaviour does not fit the behavior of the state but I don't have a good suggestion for an easy refactoring right now

tabeatheunicorn commented on angular-frontend/src/app/graph/graph-data.service.ts:
```diff
@@ -58,6 +58,11 @@ export class DataSourceService {
+  * Purpose: Provide a stable reference that can be rescaled by d3-zoom 
+  * This does NOT depend on the current zoom state  
+  */
   private readonly baseX = linkedSignal({
```
This does not add additional information over what is already written in the code

I expect sth like
/**
* Computing Basis Axis solely from current data domain and graph dimensions. 
* For actual axis, see {@linke <name of the signal>}. Here, additional alteratios from zooming etc. are taken into account
*/

AKMaily commented:
@tabeatheunicorn updated as suggested. Added the zoom functionality as Directive, moved the UnwrapSignal into a utils folder and added documentation. Also updated the doc with the latest comments in the last commit. 

Edited the description for this PR as well.
